src/lib/crypto/kimchi_backend/gadgets/ec_group.ml

```diff
@@ -0,0 +1,3965 @@
+      in
+      (* Bounds 2: Left input already checked by (Bounds 1)
+       *           Right input public parameter (no check necessary)
+       *           Result bound check below
+       *)
+      (* Add x_squared_a bound check *)
+      Foreign_field.External_checks.append_bound_check external_checks
+      @@ Foreign_field.Element.Standard.to_limbs x_squared_a ;
+      x_squared_a )
+    else x_squared
+  in
+
+  (* C3: x^3 + ax = (x^2 + a) * x *)
+  let x_cubed_ax =
+    Foreign_field.mul
+      (module Circuit)
+      external_checks x_squared_a x curve.modulus
+  in
+
+  (* Bounds 3: Left input already checked by (Bounds 2) or (Bounds 1)
+   *           Right input is gadget input so checked externally
+   *           Result bound check already tracked by external_checks
+   *)
+
+  (* C4: Optionally constrain addition of curve parameter b *)
+  let x_cubed_ax_b =
+    if not Bignum_bigint.(curve.bignum.b = zero) then (
+      (* (x^2 + a) * x + b *)
+      let x_cubed_ax_b =
+        Foreign_field.add
+          (module Circuit)
+          ~full:false x_cubed_ax curve.b curve.modulus
+      in
+      (* Result row *)
+      Foreign_field.result_row
+        (module Circuit)
+        ~label:"Ec_group.is_on_curve_x_cubed_ax_b" x_cubed_ax_b ;
+
+      (* Bounds 4: Left input already checked by (Bounds 3)
+       *           Right input public parameter (no check necessary)
+       *           Result bound check below
+       *)
+
+      (* Add x_cubed_ax_b bound check *)
+      Foreign_field.External_checks.append_bound_check external_checks
+      @@ Foreign_field.Element.Standard.to_limbs x_cubed_ax_b ;
+
+      x_cubed_ax_b )
+    else x_cubed_ax
+  in
+
+  (* C5: y^2 = y * y *)
+  let y_squared =
+    Foreign_field.mul (module Circuit) external_checks y y curve.modulus
+  in
+
+  (* Bounds 5: Left and right inputs are gadget input so checked externally
+   *           Result bound check already tracked by external_checks
+   *)
+
+  (* Copy y_squared to x_cubed_ax_b *)
+  Foreign_field.Element.Standard.assert_equal
+    (module Circuit)
+    y_squared x_cubed_ax_b ;
+  ()
+
+(* Gadget to constrain that initial accumulator (ia) point is on elliptic curve and the computation of its negation.
+ *   Note: The value of the ia itself is a deterministically generated public constant (this computation is not checked),
+ *         so using this gadget is only required in some situations.
+ *)
+let check_ia (type f) (module Circuit : Snark_intf.Run with type field = f)
+    (external_checks : f Foreign_field.External_checks.t)
+    (curve : f Curve_params.InCircuit.t) (ia : f Affine.t Curve_params.ia_points)
+    =
+  (* C1: Check that initial accumulator point is on curve *)
+  is_on_curve (module Circuit) external_checks curve ia.acc ;
+
+  (* C2: Constrain computation of the negated initial accumulator point *)
+  let neg_init_acc = negate (module Circuit) curve ia.acc in
+  (* Result row *)
+  Foreign_field.result_row
```

Comment:
Ok I see you included the result row manually, but you should warn this in the comment up front of the negate gadget.

Reply:
good idea